http-clients/src/main/java/com/palantir/remoting/http/Decorators.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ */
+
+package com.palantir.remoting.http;
+
+import com.google.common.base.Preconditions;
+import com.google.common.reflect.TypeToken;
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+import java.util.Set;
+
+/**
+ * Methods that allow classes that use the decorator pattern to be instantiated in a declarative manner.
+ */
+final class Decorators {
```

Comment:
This is a little too much magic for my liking, TBH. When composition is true delegation (i.e., the response of the delegator always depends on the response of the delegate), then IMO it's fine to use something like `new Delegator(new Delegate())`. Things got a little ugly where we are composing `new A(new B(new C()))` although A or B usually short-circuit if they can. In that case a better pattern is probably to make explicit which handler (decoder, encoder, etc) can handle which object, either by return value (Optional<Foo>) or by an explicit Boolean canHandle(...). Then a generic `return Chain.firstHandler(new A(), new B(), new C())` will return the return value of the first handler than `canHandle()` the request. This makes very explicit that the behaviour of A does not depend on the behaviour of B or C.